File: src/main/java/uk/co/alt236/tabclirun/ui/views/CommandTab.java
package uk.co.alt236.tabclirun.ui.views;

import uk.co.alt236.tabclirun.Command;
import uk.co.alt236.tabclirun.exec.Line;
import uk.co.alt236.tabclirun.exec.Result;
import uk.co.alt236.tabclirun.ui.implementations.ColorPane;

import javax.swing.*;
import java.awt.*;

class CommandTab extends JPanel {
    private static final Color COLOR_ERR = Color.decode("#E94F64");
    private static final Color COLOR_STD = Color.WHITE;
    private final JTextField textField;
    private final ColorPane textArea;
    private final FontProvider fontProvider;

    private Command command;

    CommandTab() {
        fontProvider = new FontProvider();
        textField = new JTextField();
        textArea = new ColorPane();

        setLayout(new BoxLayout(this, BoxLayout.Y_AXIS));

        setupTextArea();

        textField.setMaximumSize(new Dimension(Integer.MAX_VALUE, textField.getPreferredSize().height));
        textField.setEditable(false);

        add(textField);
        add(new JScrollPane(textArea));
    }

    void setCommand(final Command command) {
        this.command = command;

        textField.setText(command.getCommand());

        if (command.getBackgroundColor() != null) {
            textArea.setBackground(command.getBackgroundColor());
        }
    }

    void setResult(final Result result) {
        textArea.setEditable(true);

        for (final Line line : result.getLines()) {
            final Color color = line.isError()
                    ? getErrorColor()
                    : getTextColor();

            textArea.append(color, line.getText() + "\n");
        }

        textArea.setEditable(false);
    }

    private void setupTextArea() {
        Font defaultFont = UIManager.getDefaults().getFont("TextPane.font");
        int defaultSize = defaultFont.getSize();

        textArea.setFont(new Font(fontProvider.getMonospaceFontName(), Font.PLAIN, defaultSize));
        textArea.setEditable(false);
    }

    private Color getErrorColor() {
        return command.getErrorColor() == null
                ? COLOR_ERR
                : command.getErrorColor();
    }

    private Color getTextColor() {
        return command.getTextColor() == null
                ? COLOR_STD
                : command.getTextColor();
    }
}

File: app/src/main/java/uk/co/alt236/tabclirun/ui/views/tabcontent/CommandTab.kt
package uk.co.alt236.tabclirun.ui.views.tabcontent

import dev.alt236.tabclirun.libs.config.Command
import dev.alt236.tabclirun.libs.exec.result.CommandOutput
import uk.co.alt236.tabclirun.ui.implementations.ColorPane
import uk.co.alt236.tabclirun.ui.views.common.FontProvider
import java.awt.Dimension
import java.awt.Font
import javax.swing.BoxLayout
import javax.swing.JPanel
import javax.swing.JScrollPane
import javax.swing.JTextField
import javax.swing.UIManager

internal class CommandTab(verbose: Boolean) : JPanel() {
    private val commandNameField = JTextField()
    private val commandOutputField = ColorPane()
    private val fontProvider = FontProvider()
    private val resultBinder = ResultBinder(commandOutputField, verbose)

    private lateinit var command: Command
    private lateinit var colors: CommandColors


    init {
        layout = BoxLayout(this, BoxLayout.Y_AXIS)

        setupTextArea()

        commandNameField.maximumSize = Dimension(Int.MAX_VALUE, commandNameField.preferredSize.height)
        commandNameField.isEditable = false

        add(commandNameField)
        add(JScrollPane(commandOutputField))
    }

    fun setCommand(command: Command) {
        this.command = command
        this.colors = CommandColors(command)

        commandNameField.text = command.command

        if (command.backgroundColor != null) {
            commandOutputField.background = command.backgroundColor
        }
    }

    fun setResult(commandOutput: CommandOutput) {
        commandOutputField.isEditable = true

        resultBinder.bind(colors, commandOutput)

        if (command.isDisableAutoScroll && commandOutputField.text.isNotEmpty()) {
            commandOutputField.caretPosition = 0
        }

        commandOutputField.isEditable = false
    }

    private fun setupTextArea() {
        val defaultFont = UIManager.getDefaults().getFont("TextPane.font")
        val defaultSize = defaultFont.size

        commandOutputField.font = Font(fontProvider.monospaceFontName, Font.PLAIN, defaultSize)
        commandOutputField.isEditable = false
    }

}